File: fe/src/main/java/org/apache/impala/util/IcebergMetadataScanner.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.impala.util;

import com.google.common.base.Preconditions;

import java.nio.ByteBuffer;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import org.apache.iceberg.Accessor;
import org.apache.iceberg.DataTask;
import org.apache.iceberg.FileScanTask;
import org.apache.iceberg.MetadataTableType;
import org.apache.iceberg.MetadataTableUtils;
import org.apache.iceberg.StructLike;
import org.apache.impala.catalog.FeIcebergTable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.apache.iceberg.Table;
import org.apache.iceberg.TableScan;
import org.apache.iceberg.io.CloseableIterator;

/**
 * The metadata table scanner class to scan Iceberg metadata tables through the Iceberg
 * API. This object is instantiated and governed by {IcebergMetadataScanNode} at the
 * backend during scanning.
 *
 * Iceberg generally throws RuntimeExceptions, these have to be taken care of by the
 * caller of {@code IcebergMetadataScanner}.
 */
public class IcebergMetadataScanner {
  private final static Logger LOG = LoggerFactory.getLogger(IcebergMetadataScanner.class);

  // Metadata table instance.
  final private Table metadataTable_;

  // FeTable object is extracted by the backend and passed when this object is created
  final private FeIcebergTable iceTbl_;

  // Name of the metadata table
  final private String metadataTableName_;

  // Persist the file scan task iterator so we can continue after a RowBatch is full
  private CloseableIterator<FileScanTask> fileScanTaskIterator_;

  // Persist the data rows iterator, so we can continue after a batch is filled
  private CloseableIterator<StructLike> dataRowsIterator_;

  public IcebergMetadataScanner(FeIcebergTable iceTbl, String metadataTableName) {
    Preconditions.checkNotNull(iceTbl);
    this.iceTbl_ = (FeIcebergTable) iceTbl;
    this.metadataTableName_ = metadataTableName;
    this.metadataTable_ = MetadataTableUtils.createMetadataTableInstance(
      iceTbl_.getIcebergApiTable(), MetadataTableType.valueOf(metadataTableName_));
  }

  /**
   * Creates the Metadata{Table} which is a predifined Iceberg {Table} object. This method
   * also starts an Iceberg {TableScan} to scan the {Table}. After the scan is ready it
   * initializes the iterators, so the {GetNext} call can start fetching the rows through
   * the Iceberg Api.
   */
  public void ScanMetadataTable() {
    // Create and scan the metadata table
    LOG.trace("Metadata table schema: " + metadataTable_.schema().toString());
    TableScan scan = metadataTable_.newScan();
    // Init the FileScanTask iterator and DataRowsIterator
    fileScanTaskIterator_ = scan.planFiles().iterator();
    FindFileScanTaskWithRows();
  }

  /**
   * Iterates over the {{fileScanTaskIterator_}} to find a {FileScanTask} that has rows.
   */
  private boolean FindFileScanTaskWithRows() {
    while (fileScanTaskIterator_.hasNext()) {
      DataTask dataTask = (DataTask)fileScanTaskIterator_.next();
      dataRowsIterator_ = dataTask.rows().iterator();
      if (dataRowsIterator_.hasNext()) return true;
    }
    return false;
  }

  /**
   * Returns the next available row of the scan result. The row is a {StructLike} object
   * and its fields can be accessed with {Accessor}s.
   */
  public StructLike GetNext() {
    // Return the next row in the DataRows iterator
    if (dataRowsIterator_ != null && dataRowsIterator_.hasNext()) {
      return dataRowsIterator_.next();
    }
    // Otherwise this DataTask is empty, find a FileScanTask that has a non-empty DataTask
    if (FindFileScanTaskWithRows()) {
      return dataRowsIterator_.next();
    }
    return null;
  }

  /**
   * Uses the Accessor to access the value in the StructLike object. This only works for
   * non collection types.
   */
  public Object GetValueByFieldId(StructLike structLike, int fieldId) {
    Accessor accessor = metadataTable_.schema().accessorForField(fieldId);
    return accessor.get(structLike);
  }

  /**
   * Accesses the value inside the StructLike by its position.
   */
  public <T> T GetValueByPosition(StructLike structLike, int pos, Class<T> javaClass)
  {
    return structLike.get(pos, javaClass);
  }


  /**
   * Extracts the contents of a ByteBuffer into a byte array.
   */
  public byte[] ByteBufferToByteArray(ByteBuffer buffer) {
    int length = buffer.remaining();
    byte[] res = new byte[length];
    buffer.get(res);
    return res;
  }

  /**
   * Wrapper around an array or a map that is the result of a metadata table scan.
   * It is used to avoid iterating over a list through JNI.
   * For arrays the returned objects are the elements of the array, for maps they are
   * {Map.Entry} objects containing the key and value.
   */
  public static class CollectionScanner<T> {
    private Iterator<T> iterator_;

    private CollectionScanner(Iterator<T> iterator) {
      Preconditions.checkNotNull(iterator);
      iterator_ = iterator;
    }

    public static <G> CollectionScanner<G> fromArray(List<G> array) {
      CollectionScanner<G> res = new CollectionScanner<>(array.iterator());
      LOG.trace("Created metadata table array scanner, array size: " + array.size());
      return res;
    }

    public static <K, V> CollectionScanner<Map.Entry<K, V>> fromMap(Map<K, V> map) {
      CollectionScanner<Map.Entry<K, V>> res =
          new CollectionScanner<>(map.entrySet().iterator());
      LOG.trace("Created metadata table map scanner, map size: " + map.size());
      return res;
    }

    public T GetNextCollectionItem() {
      if (iterator_.hasNext()) return iterator_.next();
      return null;
    }
  }
}

File: be/src/exec/iceberg-metadata/iceberg-metadata-scanner.cc
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "exec/iceberg-metadata/iceberg-metadata-scanner.h"
#include "util/jni-util.h"

namespace impala {

IcebergMetadataScanner::IcebergMetadataScanner(jobject jtable,
    const char* metadata_table_name, const TupleDescriptor* tuple_desc)
  : jtable_(jtable),
    metadata_table_name_(metadata_table_name),
    tuple_desc_(tuple_desc) {}

Status IcebergMetadataScanner::InitJNI() {
  DCHECK(iceberg_metadata_scanner_cl_ == nullptr) << "InitJNI() already called!";
  JNIEnv* env = JniUtil::GetJNIEnv();
  if (env == nullptr) return Status("Failed to get/create JVM");

  // Global class references:
  RETURN_IF_ERROR(JniUtil::GetGlobalClassRef(env,
      "org/apache/impala/util/IcebergMetadataScanner",
      &iceberg_metadata_scanner_cl_));
  RETURN_IF_ERROR(JniUtil::GetGlobalClassRef(env,
      "org/apache/impala/util/IcebergMetadataScanner$CollectionScanner",
      &iceberg_metadata_scanner_collection_scanner_cl_));
  RETURN_IF_ERROR(JniUtil::GetGlobalClassRef(env,
      "java/util/Map$Entry", &map_entry_cl_));
  RETURN_IF_ERROR(JniUtil::GetGlobalClassRef(env, "java/util/List", &list_cl_));
  RETURN_IF_ERROR(JniUtil::GetGlobalClassRef(env, "java/util/Map", &map_cl_));
  RETURN_IF_ERROR(JniUtil::GetGlobalClassRef(env, "[B", &byte_array_cl_));

  // Method ids:
  RETURN_IF_ERROR(JniUtil::GetMethodID(env, iceberg_metadata_scanner_cl_,
      "<init>", "(Lorg/apache/impala/catalog/FeIcebergTable;Ljava/lang/String;)V",
      &iceberg_metadata_scanner_ctor_));
  RETURN_IF_ERROR(JniUtil::GetMethodID(env, iceberg_metadata_scanner_cl_,
      "ScanMetadataTable", "()V", &iceberg_metadata_scanner_scan_metadata_table_));
  RETURN_IF_ERROR(JniUtil::GetMethodID(env, iceberg_metadata_scanner_cl_,
      "GetNext", "()Lorg/apache/iceberg/StructLike;",
      &iceberg_metadata_scanner_get_next_));
  RETURN_IF_ERROR(JniUtil::GetMethodID(env, iceberg_metadata_scanner_cl_,
      "GetValueByFieldId", "(Lorg/apache/iceberg/StructLike;I)Ljava/lang/Object;",
      &iceberg_metadata_scanner_get_value_by_field_id_));
  RETURN_IF_ERROR(JniUtil::GetMethodID(env, iceberg_metadata_scanner_cl_,
      "GetValueByPosition",
      "(Lorg/apache/iceberg/StructLike;ILjava/lang/Class;)Ljava/lang/Object;",
      &iceberg_metadata_scanner_get_value_by_position_));
  RETURN_IF_ERROR(JniUtil::GetStaticMethodID(env,
      iceberg_metadata_scanner_collection_scanner_cl_, "fromArray",
      "(Ljava/util/List;)Lorg/apache/impala/util/"
      "IcebergMetadataScanner$CollectionScanner;",
      &iceberg_metadata_scanner_collection_scanner_from_array_));
  RETURN_IF_ERROR(JniUtil::GetStaticMethodID(env,
      iceberg_metadata_scanner_collection_scanner_cl_, "fromMap",
      "(Ljava/util/Map;)Lorg/apache/impala/util/"
      "IcebergMetadataScanner$CollectionScanner;",
      &iceberg_metadata_scanner_collection_scanner_from_map_));
  RETURN_IF_ERROR(JniUtil::GetMethodID(env,
      iceberg_metadata_scanner_collection_scanner_cl_,
      "GetNextCollectionItem", "()Ljava/lang/Object;",
      &iceberg_metadata_scanner_collection_scanner_get_next_collection_item_));
  RETURN_IF_ERROR(JniUtil::GetMethodID(env, iceberg_metadata_scanner_cl_,
      "ByteBufferToByteArray", "(Ljava/nio/ByteBuffer;)[B",
      &iceberg_metadata_scanner_byte_buffer_to_byte_array_));

  RETURN_IF_ERROR(JniUtil::GetMethodID(env, map_entry_cl_, "getKey",
      "()Ljava/lang/Object;", &map_entry_get_key_));
  RETURN_IF_ERROR(JniUtil::GetMethodID(env, map_entry_cl_, "getValue",
      "()Ljava/lang/Object;", &map_entry_get_value_));
  return Status::OK();
}

Status IcebergMetadataScanner::Init(JNIEnv* env) {
  jstring jstr_metadata_table_name = env->NewStringUTF(metadata_table_name_);
  jobject jmetadata_scanner = env->NewObject(iceberg_metadata_scanner_cl_,
      iceberg_metadata_scanner_ctor_, jtable_, jstr_metadata_table_name);
  RETURN_ERROR_IF_EXC(env);
  RETURN_IF_ERROR(JniUtil::LocalToGlobalRef(env, jmetadata_scanner, &jmetadata_scanner_));
  RETURN_IF_ERROR(InitSlotIdFieldIdMap(env));
  return Status::OK();
}

Status IcebergMetadataScanner::ScanMetadataTable(JNIEnv* env) {
  env->CallObjectMethod(jmetadata_scanner_,
      iceberg_metadata_scanner_scan_metadata_table_);
  RETURN_ERROR_IF_EXC(env);
  return Status::OK();
}

Status IcebergMetadataScanner::InitSlotIdFieldIdMap(JNIEnv* env) {
  for (const SlotDescriptor* slot_desc: tuple_desc_->slots()) {
    int field_id = -1;
    if (slot_desc->col_path().size() == 1) {
      // Top level slots have ColumnDescriptors that store the field ids.
      field_id = tuple_desc_->table_desc()->GetColumnDesc(slot_desc).field_id();
    } else {
      // Non top level slots are fields of a nested type. This code path is to handle
      // slots that do not have their nested type's tuple available.
      // This loop finds the struct ColumnType node that stores the slot as it has the
      // field id list of its children.
      int root_type_index = slot_desc->col_path()[0];
      const ColumnType* current_type =
          &(tuple_desc_->table_desc()->col_descs()[root_type_index].type());
      for (int i = 1; i < slot_desc->col_path().size() - 1; ++i) {
        current_type = &current_type->children[slot_desc->col_path()[i]];
      }
      field_id = current_type->field_ids[slot_desc->col_path().back()];
    }
    DCHECK_NE(field_id, -1);
    slot_id_to_field_id_map_[slot_desc->id()] = field_id;
    if (slot_desc->type().IsStructType()) {
      RETURN_IF_ERROR(InitSlotIdFieldIdMapForStruct(env, slot_desc));
    }
  }
  return Status::OK();
}

Status IcebergMetadataScanner::InitSlotIdFieldIdMapForStruct(JNIEnv* env,
    const SlotDescriptor* struct_slot_desc) {
  DCHECK(struct_slot_desc->type().IsStructType());
  const std::vector<int>& struct_field_ids = struct_slot_desc->type().field_ids;
  for (const SlotDescriptor* child_slot_desc:
      struct_slot_desc->children_tuple_descriptor()->slots()) {
    int field_id = struct_field_ids[child_slot_desc->col_path().back()];
    slot_id_to_field_id_map_[child_slot_desc->id()] = field_id;
    if (child_slot_desc->type().IsStructType()) {
      RETURN_IF_ERROR(InitSlotIdFieldIdMapForStruct(env, child_slot_desc));
    }
  }
  return Status::OK();
}

Status IcebergMetadataScanner::GetNext(JNIEnv* env, jobject* result) {
  *result = env->CallObjectMethod(jmetadata_scanner_, iceberg_metadata_scanner_get_next_);
  RETURN_ERROR_IF_EXC(env);
  return Status::OK();
}

Status IcebergMetadataScanner::GetNextArrayItem(JNIEnv* env, const jobject& scanner,
    jobject* result) {
  return GetNextCollectionScannerItem(env, scanner, result);
}

Status IcebergMetadataScanner::GetNextMapKeyAndValue(JNIEnv* env, const jobject& scanner,
    jobject* key, jobject* value) {
  jobject map_entry;
  RETURN_IF_ERROR(GetNextCollectionScannerItem(env, scanner, &map_entry));
  DCHECK(env->IsInstanceOf(map_entry, map_entry_cl_) == JNI_TRUE);

  *key = env->CallObjectMethod(map_entry, map_entry_get_key_);
  RETURN_ERROR_IF_EXC(env);

  *value = env->CallObjectMethod(map_entry, map_entry_get_value_);
  RETURN_ERROR_IF_EXC(env);
  env->DeleteLocalRef(map_entry);
  return Status::OK();
}

Status IcebergMetadataScanner::GetNextCollectionScannerItem(JNIEnv* env,
    const jobject& scanner, jobject* result) {
  *result = env->CallObjectMethod(scanner,
      iceberg_metadata_scanner_collection_scanner_get_next_collection_item_);
  RETURN_ERROR_IF_EXC(env);
  return Status::OK();
}


Status IcebergMetadataScanner::GetValue(JNIEnv* env, const SlotDescriptor* slot_desc,
    const jobject &struct_like_row, const jclass& clazz, jobject* result) {
  DCHECK(slot_desc != nullptr);
  auto field_id_it = slot_id_to_field_id_map_.find(slot_desc->id());
  if (field_id_it != slot_id_to_field_id_map_.end()) {
    // Use accessor when it is available, these are top level primitive types, top level
    // structs and structs inside structs.
    RETURN_IF_ERROR(GetValueByFieldId(env, struct_like_row, field_id_it->second, result));
  } else {
    // Accessor is not available, this must be a STRUCT inside an ARRAY.
    DCHECK(slot_desc->parent()->isTupleOfStructSlot());
    int pos = slot_desc->col_path().back();
    RETURN_IF_ERROR(GetValueByPosition(env, struct_like_row, pos, clazz, result));
  }

  return Status::OK();
}

Status IcebergMetadataScanner::GetValueByFieldId(JNIEnv* env, const jobject &struct_like,
    int field_id, jobject* result) {
  *result = env->CallObjectMethod(jmetadata_scanner_,
      iceberg_metadata_scanner_get_value_by_field_id_, struct_like, field_id);
  RETURN_ERROR_IF_EXC(env);
  return Status::OK();
}

Status IcebergMetadataScanner::GetValueByPosition(JNIEnv* env, const jobject &struct_like,
    int pos, const jclass &clazz, jobject* result) {
  *result = env->CallObjectMethod(jmetadata_scanner_,
      iceberg_metadata_scanner_get_value_by_position_, struct_like, pos, clazz);
  RETURN_ERROR_IF_EXC(env);
  return Status::OK();
}

Status IcebergMetadataScanner::CreateArrayScanner(JNIEnv* env, const jobject &list,
    jobject* result) {
  return CreateArrayOrMapScanner</*IS_ARRAY*/ true>(env, list, result);
}

Status IcebergMetadataScanner::CreateMapScanner(JNIEnv* env, const jobject &map,
    jobject* result) {
  return CreateArrayOrMapScanner</*IS_ARRAY*/ false>(env, map, result);
}

template <bool IS_ARRAY>
Status IcebergMetadataScanner::CreateArrayOrMapScanner(JNIEnv* env,
    const jobject &list_or_map, jobject* result) {
  DCHECK(result != nullptr);

  jmethodID* factory_method;
  if constexpr (IS_ARRAY) {
    DCHECK(env->IsInstanceOf(list_or_map, list_cl_) == JNI_TRUE);
    factory_method = &iceberg_metadata_scanner_collection_scanner_from_array_;
  } else {
    DCHECK(env->IsInstanceOf(list_or_map, map_cl_) == JNI_TRUE);
    factory_method = &iceberg_metadata_scanner_collection_scanner_from_map_;
  }

  *result = env->CallStaticObjectMethod(iceberg_metadata_scanner_collection_scanner_cl_,
      *factory_method, list_or_map);
  RETURN_ERROR_IF_EXC(env);
  return Status::OK();
}

Status IcebergMetadataScanner::ConvertJavaByteBufferToByteArray(JNIEnv* env,
    const jobject& byte_buffer, jbyteArray* result) {
  jobject arr = env->CallObjectMethod(jmetadata_scanner_,
      iceberg_metadata_scanner_byte_buffer_to_byte_array_, byte_buffer);
  RETURN_ERROR_IF_EXC(env);
  DCHECK(env->IsInstanceOf(arr, byte_array_cl_));
  *result = static_cast<jbyteArray>(arr);
  return Status::OK();
}

void IcebergMetadataScanner::Close(RuntimeState* state) {
  JNIEnv* env = JniUtil::GetJNIEnv();
  if (env != nullptr) {
    if (jmetadata_scanner_ != nullptr) env->DeleteGlobalRef(jmetadata_scanner_);
  }
}

string IcebergMetadataScanner::DebugString() {
  std::stringstream out;
  out << "IcebergMetadataScanner: [ Metadata table name: "
      << metadata_table_name_ << "; ";
  std::unordered_map<impala::SlotId, int>::iterator it = slot_id_to_field_id_map_.begin();
  if (it != slot_id_to_field_id_map_.end()) {
    out << "SlotId to FieldId map: [" << it->first << " : " << it->second;
    for (it++; it != slot_id_to_field_id_map_.end(); it++) {
      out << ", " << it->first << " : " << it->second;
    }
    out << "]; ";
  }
  out << tuple_desc_->DebugString() << "]";
  return out.str();
}

}
